File: python/grade-school/grade_school.py
class School:
    def __init__(self):
        # Initialize an empty dictionary to store the school's roster.
        # Initialize an empty list to keep track of whether students are added.
        self.school = {}
        self.add = []

    def add_student(self, name, grade):
        # Check if the student's name is not already in the roster.
        if name not in self.roster():
            # If not, add the student to the specified grade.
            # If the grade doesn't exist, create it.
            self.school[grade] = self.school.get(grade, []) + [name]
            # Append True to the 'add' list to indicate the student was added.
            self.add.append(True)
        else:
            # If the student's name is already in the roster, append False to 'add'.
            self.add.append(False)

    def added(self):
        # Return the 'add' list, which contains True for added students and False for existing ones.
        return self.add

    def roster(self):
        # Create a sorted list of student names by iterating through grades and names in each grade.
        return [name
                for grade in sorted(self.school.keys())
                for name in sorted(self.school[grade])]

    def grade(self, grade_number):
        # Return a sorted list of students in the specified grade or an empty list if the grade doesn't exist.
        return sorted(self.school.get(grade_number, []))

File: javascript/grade-school/grade-school.js
export class GradeSchool {
  constructor() {
    this.students = {}; // Initialize an empty object to store student data
  }

  // Return a deep copy of the roster, organized by grade
  roster() {
    const roster = {};

    // Iterate through the students object and create a copy of the roster
    for (const grade in this.students) {
      roster[grade] = [...this.students[grade]];
    }

    return roster;
  }

  // Add a student to a specific grade
  add(name, grade) {
    // Remove the student from all other grades if they exist
    for (const g in this.students) {
      if (this.students[g].includes(name)) {
        this.students[g] = this.students[g].filter((n) => n !== name);
      }
    }

    if (!this.students[grade]) {
      this.students[grade] = []; // Initialize an array for the grade if it doesn't exist
    }

    this.students[grade].push(name); // Add the student to the specified grade
    this.students[grade].sort(); // Sort the students' names in the grade
  }

  // Return a list of students for a specific grade
  grade(grade) {
    return this.students[grade] ? [...this.students[grade]] : []; // Return a copy of the students in the grade
  }
}
